daliuge-engine/dlg/apps/mpi.py

```diff
@@ -66,33 +76,105 @@ class MPIApp(BarrierAppDROP):
         self._use_wrapper = self._popArg(kwargs, "use_wrapper", False)
-        self._args = self._popArg(kwargs, "args", [])
+        self._args = self._popArg(kwargs, "args", "")
+        self._applicationArgs = self._popArg(kwargs, "applicationArgs", {})
+        self._argumentPrefix = self._popArg(kwargs, "argumentPrefix", "--")
+        self._paramValueSeparator = self._popArg(kwargs, "paramValueSeparator", " ")
+        self._inputRedirect = self._popArg(kwargs, "input_redirection", "")
+        self._outputRedirect = self._popArg(kwargs, "output_redirection", "")
+
+        self._command = self._popArg(kwargs, "command", None)
         if not self._command:
             raise InvalidDropException(
                 self, "No command specified, cannot create MPIApp"
             )
+        self._recompute_data = {}
 
     def run(self):
         from mpi4py import MPI
 
         cmd, args = self._command, self._args
+        inputs = self._inputs
+        outputs = self._outputs
+
+        logger.debug("Parameters found: %s", json.dumps(self.parameters))
+        logger.debug("MPI Inputs: %s; MPI Outputs: %s", inputs, outputs)
+        # we only support passing a path for bash apps
+        fsInputs = {uid: i for uid, i in inputs.items() if droputils.has_path(i)}
+        fsOutputs = {uid: o for uid, o in outputs.items() if droputils.has_path(o)}
+        dataURLInputs = {
+            uid: i for uid, i in inputs.items() if not droputils.has_path(i)
+        }
+        dataURLOutputs = {
+            uid: o for uid, o in outputs.items() if not droputils.has_path(o)
+        }
+        # deal with named ports
+        inport_names = self.parameters["inputs"] if "inputs" in self.parameters else []
+        outport_names = (
+            self.parameters["outputs"] if "outputs" in self.parameters else []
+        )
+        reader = get_port_reader_function(self.input_parser)
+        keyargs, pargs = replace_named_ports(
+            inputs.items(),
+            outputs.items(),
+            inport_names,
+            outport_names,
+            self._applicationArgs,
+            argumentPrefix=self._argumentPrefix,
+            separator=self._paramValueSeparator,
+            parser=reader,
+        )
+        argumentString = (
+            f"{' '.join(map(str,pargs + keyargs))}"  # add kwargs to end of pargs
+        )
+        # complete command including all additional parameters and optional redirects
+        if len(argumentString.strip()) > 0:
+            # the _cmdLineArgs would very likely make the command line invalid
+            cmd = f"{self._command} {argumentString} "
+        else:
+            cmd = f"{self._command} {argumentString} {args} "
+        if self._outputRedirect:
+            cmd = f"{cmd} > {self._outputRedirect}"
+        if self._inputRedirect:
+            cmd = f"cat {self._inputRedirect} > {cmd}"
+        cmd = cmd.strip()
+
+        app_uid = self.uid
```

Comment:
We don't use `app_uid` anywhere; also, I am not sure what the difference between `self.uid` is and `self._uid` is; for internal methods, I'd advocate being consistent with `self._uid`.